File: racing-writes/racing-writes.py
#! /usr/bin/env python3

import threading

INCREMENTS_PER_THREAD = 100000
THREADS = 10


# Define a small class so that we can pass a counter by reference.
class Counter:
    def __init__(self):
        self.count = 0


# In Python, as in most languages, the easy way to increment a shared counter
# is prone to race conditions. Another thread can swoop in *after* we've read
# the previous value of count, but *before* we've written the new value. The
# result is that we overwrite the other thread's work, and the counter winds up
# smaller than expected.
def bad_increment(counter):
    counter.count += 1


counter = Counter()
threads = []
for _ in range(THREADS):
    def thread_inner():
        for _ in range(INCREMENTS_PER_THREAD):
            bad_increment(counter)
    thread = threading.Thread(target=thread_inner)
    threads.append(thread)
    thread.start()
for thread in threads:
    thread.join()
print("Expected:", THREADS * INCREMENTS_PER_THREAD)
print("Actual:", counter.count)

File: racing-writes/racing-writes.rs
use std::sync::{Arc, Mutex};
use std::thread;

const INCREMENTS_PER_THREAD: u32 = 1000;
const THREADS: u32 = 10;

// Here's the safe way to increment a shared counter in Rust. Of course we
// could've used a Mutex to fix the race condition in Python too. The key
// difference here is that Rust *forces* us to use this Mutex. An unsynchronized
// implementation won't even compile!
#[allow(dead_code)]
fn good_increment(counter: Arc<Mutex<u32>>) {
    *counter.lock().unwrap() += 1;
}

// If you jump through hoops, you can make Rust reproduce the race condition
// from Python. You have to take the lock, copy the counter, *give up* the lock,
// take it *again*, and finally write the incremented value.
fn bad_increment(counter: Arc<Mutex<u32>>) {
    let value = *counter.lock().unwrap();
    // The MutexGuard above isn't locally bound, so it's immediately dropped,
    // and we're free to take it again...for SCIENCE.
    *counter.lock().unwrap() = value + 1;
}

fn main() {
    let counter = Arc::new(Mutex::new(0));
    let mut handles = Vec::new();
    for _ in 0..THREADS {
        let counter_ref = counter.clone();
        let handle = thread::spawn(move || {
            for _ in 0..INCREMENTS_PER_THREAD {
                bad_increment(counter_ref.clone());
            }
        });
        handles.push(handle);
    }
    for handle in handles {
        handle.join().unwrap();
    }
    println!("Expected: {}", THREADS * INCREMENTS_PER_THREAD);
    println!("Actual: {}", *counter.lock().unwrap());
}
